### ISSUE TITLE
Arm64 part of multi-arch images is not served/cached

### ISSUE THREAD

**Issue by gmykhailiuta** [enix/kube-image-keeper#104]
Currently the multi-arch images fail to start on arm64 nodes when Kuik is enabled. Deeper investigation revealed that local docker registry does not contain and/or serve the arm64 layers of multi-arch images like `library/docker`.

Cached images manifest (note that there's only amd64 architecture):
```bash
docker pull kube-image-keeper-registry.dev-kube-image-keeper.svc.cluster.local:5000/docker.io/library/docker:20-dind
docker manifest inspect --verbose kube-image-keeper-registry.dev-kube-image-keeper.svc.cluster.local:5000/docker.io/library/docker:20-dind --insecure
```
```json
{
	"Ref": "kube-image-keeper-registry.dev-kube-image-keeper.svc.cluster.local:5000/docker.io/library/docker:20-dind",
	"Descriptor": {
		"mediaType": "application/vnd.docker.distribution.manifest.v2+json",
		"digest": "sha256:11a8556b63283fb6edb98fc990166476bc2e14b33164039ab6132b27d84882d8",
		"size": 3251,
		"platform": {
			"architecture": "amd64",
			"os": "linux"
		}
	},
	"SchemaV2Manifest": {
		"schemaVersion": 2,
		"mediaType": "application/vnd.docker.distribution.manifest.v2+json",
		"config": {
			"mediaType": "application/vnd.docker.container.image.v1+json",
			"size": 11319,
			"digest": "sha256:c0e053541b0eec2ee002f3862daedf169d40ff5e65d2a11206ebb1dd883acc13"
		},
		"layers": [
			{
				"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
				"size": 3397490,
				"digest": "sha256:8a49fdb3b6a5ff2bd8ec6a86c05b2922a0f7454579ecc07637e94dfd1d0639b6"
			},
			{
				"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
				"size": 2014629,
				"digest": "sha256:db1d8fde5ab00eb91a63f1f2ceb21e27bf1b3e1bcb622bb9a3828478273df8cb"
			},
			{
				"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
				"size": 90,
				"digest": "sha256:f3759a44eb9f3e3577a539da15a245f36a7436c3bcc726a0a4c0e677c835a6ba"
			},
			{
				"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
				"size": 14117997,
				"digest": "sha256:1465fd8ca4ab3bbf2ff1fd0a79ecbb543f6c91c45c4e6643f12bf8f172af186b"
			},
			{
				"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
				"size": 16001759,
				"digest": "sha256:95c1bc752a8b6c0d7c1951d978f7a9d88f0473002f2b6fe9c05052dfa1d64144"
			},
			{
				"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
				"size": 16384833,
				"digest": "sha256:f5ce33fb9a21fd7c58a485b4048bb20b38c48518028ebbc87b46da23e62cd48e"
			},
			{
				"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
				"size": 548,
				"digest": "sha256:4d37eec86745434c6e5165a35a2012abbfd788ef455e7a6f3aeb282e2ef4839e"
			},
			{
				"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
				"size": 1020,
				"digest": "sha256:49d0e23604c0500f6a93a56a7e0951c8b8264e5822db51d8b4105d6c1a373374"
			},
			{
				"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
				"size": 167,
				"digest": "sha256:ac56a29f30eb929bd39086c9e11300ca62e2e446ae9d5d557e6723ded96d1ccc"
			},
			{
				"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
				"size": 7025508,
				"digest": "sha256:0184c8e99504ad3ef831f05b75e81580f06a9620c9b41627c5d92825baee5dcf"
			},
			{
				"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
				"size": 1319,
				"digest": "sha256:23430c064101ef52f3b552e39750e61826d539d4e98b43d88c805e504a591291"
			},
			{
				"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
				"size": 53901045,
				"digest": "sha256:e2a9b00eaf87e56919b17afcd93f6862875ce653ab4de6d7a3d5c916ef833944"
			},
			{
				"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
				"size": 1048,
				"digest": "sha256:004faac3bc8662333a4426e89bb9524953842efa0d3355697d3e5961fd2abeba"
			},
			{
				"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
				"size": 2814,
				"digest": "sha256:765a9f2e07176c06faf302ae4fdcbf43e75986f2fed0d7d0ca0973cc54276225"
			}
		]
	}
}
```

Original image manifest served by dockerhub (both amd64 and arm64 are present):
```bash
docker image pull library/docker:20-dind
docker manifest inspect --verbose library/docker:20-dind
```
```json
[
	{
		"Ref": "docker.io/library/docker:20-dind@sha256:11a8556b63283fb6edb98fc990166476bc2e14b33164039ab6132b27d84882d8",
		"Descriptor": {
			"mediaType": "application/vnd.docker.distribution.manifest.v2+json",
			"digest": "sha256:11a8556b63283fb6edb98fc990166476bc2e14b33164039ab6132b27d84882d8",
			"size": 3251,
			"platform": {
				"architecture": "amd64",
				"os": "linux"
			}
		},
		"SchemaV2Manifest": {
			"schemaVersion": 2,
			"mediaType": "application/vnd.docker.distribution.manifest.v2+json",
			"config": {
				"mediaType": "application/vnd.docker.container.image.v1+json",
				"size": 11319,
				"digest": "sha256:c0e053541b0eec2ee002f3862daedf169d40ff5e65d2a11206ebb1dd883acc13"
			},
			"layers": [
				{
					"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
					"size": 3397490,
					"digest": "sha256:8a49fdb3b6a5ff2bd8ec6a86c05b2922a0f7454579ecc07637e94dfd1d0639b6"
				},
				{
					"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
					"size": 2014629,
					"digest": "sha256:db1d8fde5ab00eb91a63f1f2ceb21e27bf1b3e1bcb622bb9a3828478273df8cb"
				},
				{
					"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
					"size": 90,
					"digest": "sha256:f3759a44eb9f3e3577a539da15a245f36a7436c3bcc726a0a4c0e677c835a6ba"
				},
				{
					"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
					"size": 14117997,
					"digest": "sha256:1465fd8ca4ab3bbf2ff1fd0a79ecbb543f6c91c45c4e6643f12bf8f172af186b"
				},
				{
					"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
					"size": 16001759,
					"digest": "sha256:95c1bc752a8b6c0d7c1951d978f7a9d88f0473002f2b6fe9c05052dfa1d64144"
				},
				{
					"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
					"size": 16384833,
					"digest": "sha256:f5ce33fb9a21fd7c58a485b4048bb20b38c48518028ebbc87b46da23e62cd48e"
				},
				{
					"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
					"size": 548,
					"digest": "sha256:4d37eec86745434c6e5165a35a2012abbfd788ef455e7a6f3aeb282e2ef4839e"
				},
				{
					"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
					"size": 1020,
					"digest": "sha256:49d0e23604c0500f6a93a56a7e0951c8b8264e5822db51d8b4105d6c1a373374"
				},
				{
					"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
					"size": 167,
					"digest": "sha256:ac56a29f30eb929bd39086c9e11300ca62e2e446ae9d5d557e6723ded96d1ccc"
				},
				{
					"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
					"size": 7025508,
					"digest": "sha256:0184c8e99504ad3ef831f05b75e81580f06a9620c9b41627c5d92825baee5dcf"
				},
				{
					"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
					"size": 1319,
					"digest": "sha256:23430c064101ef52f3b552e39750e61826d539d4e98b43d88c805e504a591291"
				},
				{
					"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
					"size": 53901045,
					"digest": "sha256:e2a9b00eaf87e56919b17afcd93f6862875ce653ab4de6d7a3d5c916ef833944"
				},
				{
					"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
					"size": 1048,
					"digest": "sha256:004faac3bc8662333a4426e89bb9524953842efa0d3355697d3e5961fd2abeba"
				},
				{
					"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
					"size": 2814,
					"digest": "sha256:765a9f2e07176c06faf302ae4fdcbf43e75986f2fed0d7d0ca0973cc54276225"
				}
			]
		}
	},
	{
		"Ref": "docker.io/library/docker:20-dind@sha256:f52db26a8460b6e4da8ff56ce235570fcb81fcef685bf1aebaa0290da6494599",
		"Descriptor": {
			"mediaType": "application/vnd.docker.distribution.manifest.v2+json",
			"digest": "sha256:f52db26a8460b6e4da8ff56ce235570fcb81fcef685bf1aebaa0290da6494599",
			"size": 3251,
			"platform": {
				"architecture": "arm64",
				"os": "linux",
				"variant": "v8"
			}
		},
		"SchemaV2Manifest": {
			"schemaVersion": 2,
			"mediaType": "application/vnd.docker.distribution.manifest.v2+json",
			"config": {
				"mediaType": "application/vnd.docker.container.image.v1+json",
				"size": 11334,
				"digest": "sha256:3abfbe5cf820b38df0a10d53261bc2b376ec5ff6680795a4b3eee1e70aed937d"
			},
			"layers": [
				{
					"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
					"size": 3342848,
					"digest": "sha256:08409d4172603f40b56eb6b76240a1e6bd78baa0e96590dc7ff76c5f1a093af2"
				},
				{
					"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
					"size": 2025067,
					"digest": "sha256:00b32f8d4d5f789274c3df956d1a71dbda3c0d8eb267b33c53aa6de02b3e49fc"
				},
				{
					"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
					"size": 92,
					"digest": "sha256:249b647e77d07a61a6a56de29194440506db84b546bd7ad42c73b8f4388a30cd"
				},
				{
					"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
					"size": 12836224,
					"digest": "sha256:70115aff51274d0d1ff79c0b64a2364768dafee3c4f858d14c65e1653fea4259"
				},
				{
					"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
					"size": 14441515,
					"digest": "sha256:75265b819afd64316c90682b8f16ecb72b9a828eb3535892f735cfe778ad9ad1"
				},
				{
					"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
					"size": 14835063,
					"digest": "sha256:7e3009834ea7bfadd539c5f6f8679ff926b95c69041b9490d15d1f5584dccf3f"
				},
				{
					"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
					"size": 545,
					"digest": "sha256:ceae6ddcdf8c2486b41a2d8c467867d68a46d3dc4cca0b8ee3d1e81b68b87a6a"
				},
				{
					"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
					"size": 1019,
					"digest": "sha256:bf932719bc1435f4e06003bae3823a2eebb71ad2c5e2e5e0327a1f2ce0ac2d29"
				},
				{
					"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
					"size": 166,
					"digest": "sha256:8b6a26a6efce5f5993747f0eaf938c5ba0ad4989ccdc527d783296447186b423"
				},
				{
					"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
					"size": 7245237,
					"digest": "sha256:40af53c00db16857c71785cabd9051fe6a84accfd8877f9ace9769152a95c0ec"
				},
				{
					"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
					"size": 1322,
					"digest": "sha256:620f16fecdb17140029feeda0f58e4553c89f382e3f27731fb0ca33cd7d4ee51"
				},
				{
					"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
					"size": 49320282,
					"digest": "sha256:bddcb496053fdf47ac8750b107cc14ea095ae5957ba8eeec32ec3d5f4857be12"
				},
				{
					"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
					"size": 1050,
					"digest": "sha256:f14a1a3a1065ccb47908b268e7f5117a9250523b2c989888a893b21c94040496"
				},
				{
					"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
					"size": 2815,
					"digest": "sha256:fc5829b166b80d4ae63da17dd529566392c5237f1496f3f9998e844d08d02c69"
				}
			]
		}
	}
]
```
The pod's spec that was used to obtain the output above:
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: dind
  labels:
    app: dind
    kube-image-keeper.enix.io/image-caching-policy: "ignore"
spec:
  containers:
    - image: docker:20-dind
      command:
        - "sleep"
        - "604800"
      imagePullPolicy: IfNotPresent
      name: dind
      env:
        # - name: DOCKER_HOST
        #   value: tcp://localhost:2375
        - name: DOCKER_BUILDKIT
          value: "1"
        - name: DOCKER_DRIVER
          value: overlay2
        - name: DOCKER_TLS_CERTDIR
          value: ""
      volumeMounts:
        - name: dind-storage
          mountPath: /var/lib/docker
        - name: docker-socket
          mountPath: /var/run
    - name: docker
      image: docker:20-dind
      securityContext:
        privileged: true
      args:
        - "--tls=false"
        - "--insecure-registry=kube-image-keeper-registry.dev-kube-image-keeper.svc.cluster.local:5000"
      volumeMounts:
        - name: dind-storage
          mountPath: /var/lib/docker
        - name: docker-socket
          mountPath: /var/run
  nodeSelector:
    kubernetes.io/os: "linux"
    kubernetes.io/arch: "arm64"
  volumes:
    - name: dind-storage
      emptyDir: {}
    - name: docker-socket
      emptyDir: {}
```

**Comment by Nicolasgouze** [comment]
Sorry for the delay @gmykhailiuta, this ticket is part of our next prio !

**Comment by OGKevin** [comment]
I can conform that the fix provided in #119 seems to work fine 👌🏾
